crates/bevy_reflect/derive/src/enum_utility.rs

```diff
@@ -1,122 +1,263 @@
-            let field_ident = ident_or_index(field.data.ident.as_ref(), declare_index);
-            let field_ty = &field.data.ty;
-
-            let field_value = if field.attrs.ignore.is_ignored() {
-                match &field.attrs.default {
-                    DefaultBehavior::Func(path) => quote! { #path() },
-                    _ => quote! { #FQDefault::default() }
-                }
-            } else {
-                let field_accessor = match &field.data.ident {
-                    Some(ident) => {
-                        let name = ident.to_string();
-                        quote!(#ref_value.field(#name))
-                    }
-                    None => quote!(#ref_value.field_at(#reflect_index)),
-                };
-                reflect_index += 1;
-
-                let (resolve_error, resolve_missing) = if return_apply_error {
-                    let field_ref_str = match &field_ident {
-                        Member::Named(ident) => format!("{ident}"),
-                        Member::Unnamed(index) => format!(".{}", index.index)
-                    };
-                    let ty = field.data.ty.to_token_stream();
-
-                    (
-                        quote!(.ok_or(#bevy_reflect_path::ApplyError::MismatchedTypes {
-                            // The unwrap won't panic. By this point the #field_accessor would have been invoked once and any failure to
-                            // access the given field handled by the `resolve_missing` code bellow.
-                            from_type: ::core::convert::Into::into(
-                                #bevy_reflect_path::DynamicTypePath::reflect_type_path(#FQOption::unwrap(#field_accessor))
-                            ),
-                            to_type: ::core::convert::Into::into(<#ty as #bevy_reflect_path::TypePath>::type_path())
-                        })?),
-                        quote!(.ok_or(#bevy_reflect_path::ApplyError::MissingEnumField {
-                                variant_name: ::core::convert::Into::into(#name),
-                                field_name: ::core::convert::Into::into(#field_ref_str)
-                        })?)
-                    )
-                } else {
-                    (quote!(?), quote!(?))
+#[derive(Copy, Clone)]
+struct VariantField<'a, 'b> {
+    /// The pre-computed member for the field.
+    pub member: &'a Member,
+    /// The name of the variant that contains the field.
+    pub variant_name: &'a str,
+    /// The field data.
+    pub field: &'a StructField<'b>,
+}
+
+/// A builder for constructing [`EnumVariantOutputData`].
+///
+/// When a variant field needs to be constructed, the builder will:
+/// 1. Use `field_accessor` to access the field as an `Option<&dyn Reflect>`.
+/// 2. Use `field_unwrapper` to unwrap the field into a `&dyn Reflect`.
+/// 3. Use `field_constructor` to construct the field into a concrete value.
+struct EnumVariantOutputDataBuilder<'a> {
```

Comment:
I'm not sure if this design with `field_accessor`, `field_unwrapper`, `field_constructor` is the clearest.

Maybe just having a `BaseEnumVariantOutputDataBuilder`
and then two different structs s`EnumVariantOutputDataBuilderForFromReflect`
and `EnumVariantOutputDataBuilderForTryApply` that share the underlying `BaseEnum...` but have different implementations for the `unwrapper` and `constructor` would be easier to understand?

I don't think it's a blocker, because you're the one primarily maintaining this code so it's already an improvement if it's clearer to you. I think the new structure is overall much clearer.

Reply:
> Maybe just having a `BaseEnumVariantOutputDataBuilder`
> and then two different structs s`EnumVariantOutputDataBuilderForFromReflect`
> and `EnumVariantOutputDataBuilderForTryApply` that share the underlying `BaseEnum...` but have different implementations for the `unwrapper` and `constructor` would be easier to understand?

Hm, could you give an example? I'm probably misunderstanding, but this almost sounds like what I have now with the base builder type and the two `pub(crate)` generator functions.

Reply:
Hm I guess you're right.
I was thinking of something with generics 
```
trait EnumTokenBuilder {
   fn field_accessor(&Ident, VariantField) -> TokenStream;
   
   fn ...
}
```
just to avoid the `Box<dyn Fn>` and having to build the struct without those functions existing.
I guess I'm just not a fan of the builder pattern in this instance, especially if the fields are not optional. A new contributor might not understand why the `new` function doesn't just require those fields.
It's mostly a matter of taste, though

Reply:
Yeah tbh I don't love it either. Alternatively, I could scrap both `new` and the `with_***` and just have it all defined like:

```rust
let builder = EnumVariantOutputDataBuilder {
  this,
  reflect_enum,
  field_accessor: Box::new(/* ... */),
  // ...
};

builder.build()
```

It just means that any fields with a potential default value will need to be defined each time.

The trait method you propose is nice too and is probably a bit Rustier. I might look into that actually. 🤔

Reply:
I've updated the code to use traits instead. I think it's a lot cleaner so thank you for the suggestion!

Let me know if you have any further thoughts on it (or if you even prefer the old version haha)!